### PR TITLE
Improve cache context

### PR DESCRIPTION
**Pull request by wolfogre**
Related to: #22294 #23186 #23054

Replace: #23218

Some discussion is in the comments of #23218.

Highlights:
- Add Expiration for cache context. If a cache context has been used for more than 10s, the cache data will be ignored, and warning logs will be printed.
- Add `discard` field to `cacheContext`, a `cacheContext` with `discard` true will drop all cached data and won't store any new one.
- Introduce `WithNoCacheContext`, if one wants to run long-life tasks, but the parent context is a cache context, `WithNoCacheContext(perentCtx)` will discard the cache data, so it will be safe to keep the context for a long time.
- It will be fine to treat an original context as a cache context, like `GetContextData(context.Backgraud())`, no warning logs will be printed.

Some cases about nesting:

When:
- *A*, *B* or *C* means a cache context.
- ~*A*~, ~*B*~ or ~*C*~ means a discard cache context.
- `ctx` means `context.Backgrand()`
- *A(ctx)* means a cache context with `ctx` as the parent context.
- *B(A(ctx))* means a cache context with `A(ctx)` as the parent context.
- `With` means `WithCacheContext`
- `WithNo` means `WithNoCacheContext`

So:
- `With(ctx)` -> *A(ctx)*
- `With(With(ctx))` -> *A(ctx)*, not *B(A(ctx))*
- `With(With(With(ctx)))` -> *A(ctx)*, not *C(B(A(ctx)))*
- `WithNo(ctx)` -> *ctx*, not *~A~(ctx)*
- `WithNo(With(ctx))` -> *~A~(ctx)*
- `WithNo(WithNo(With(ctx)))` -> *~A~(ctx)*, not *~B~(~A~(ctx))*
- `With(WithNo(With(ctx)))` -> *B(~A~(ctx))*
- `WithNo(With(WithNo(With(ctx))))` -> *~B~(~A~(ctx))*
- `With(WithNo(With(WithNo(With(ctx)))))` -> *C(~B~(~A~(ctx)))*

### CODE REVIEW

File: modules/cache/context.go [hunk 1]
```diff
@@ -14,65 +15,118 @@ import (
...
+	data    map[any]map[any]any
+	lock    sync.RWMutex
+	created time.Time
+	discard bool
 }
 
 func (cc *cacheContext) Get(tp, key any) any {
 	cc.lock.RLock()
 	defer cc.lock.RUnlock()
-	if cc.data[tp] == nil {
```
Comment (lunny) [inline modules/cache/context.go:25]: Why removed these lines?

Comment (wolfogre) [reply modules/cache/context.go:25]: Why not? It's safe to read or delete keys from a nil map.

**Comment by wxiaoguang** [comment]
The description seems very valuable. Maybe the description could be put into code as package comment? If would help future developers when they are using the package.

File: modules/cache/context.go [hunk 2]
```diff
@@ -14,65 +15,118 @@ import (
 // This is useful for caching data that is expensive to calculate and is likely to be
 // used multiple times in a request.
 type cacheContext struct {
-	ctx  context.Context
```
Comment (delvh) [inline modules/cache/context.go:17]: We don't store a parent ctx anymore?
Why?

Comment (wolfogre) [reply modules/cache/context.go:17]: Because it's never been used. Actually, `cacheContext` doesn't implement `context.Context`

**Review by delvh — APPROVED** [review]

File: modules/cache/context.go [hunk 3]
```diff
@@ -14,65 +15,145 @@ import (
...
+- WithNo(With(ctx)) -> A'(ctx)
+- WithNo(WithNo(With(ctx))) -> A'(ctx), not B'(A'(ctx)), don't create new cache context if we don't have to.
+- With(WithNo(With(ctx))) -> B(A'(ctx)), not A(ctx), never reuse a discard cache context.
+- WithNo(With(WithNo(With(ctx)))) -> B'(A'(ctx))
+- With(WithNo(With(WithNo(With(ctx))))) -> C(B'(A'(ctx))), so there's always only one not-discard cache context.
+*/
+
 func WithCacheContext(ctx context.Context) context.Context {
+	if c, ok := ctx.Value(cacheContextKey).(*cacheContext); ok {
+		if !c.discard {
```
Comment (lunny) [inline modules/cache/context.go]: We need r lock here.

File: modules/cache/context.go [hunk 4]
```diff
@@ -14,65 +15,145 @@ import (
...
+			// reuse parent context
+			return ctx
+		}
```
Comment (lunny) [inline modules/cache/context.go:110]: How to handle else?

Comment (wolfogre) [reply modules/cache/context.go:110]: Ignore the discard cache and create a new one. 

`With(WithNo(With(ctx)))` -> *B(~A~(ctx))*

**Review by lunny — APPROVED** [review]

**Comment by jolheiser** [comment]
🎺 🤖

**Pull request merged**